### ISSUE TITLE
getByRole fails with "TypeError: Cannot read property 'parentElement' of null" when using Canvas Kit React button with an icon

### ISSUE THREAD

**Issue by ahayes91** [testing-library/dom-testing-library#1101]
- `@testing-library/react` version: 12.1.2
- `@workday/canvas-kit-react` version "^6.3.3"
-  `@workday/canvas-system-icons-web` version "^2.0.6"
- Testing Framework and version: jest 26.6.3 (but that doesn't seem to matter, this fails on the template Codesandbox too)
  
- DOM Environment: default jsdom, running Codesandbox tests in Chrome Version 97.0.4692.71 (Official Build) (x86_64)
  

### Relevant code or config:

```js
// CanvasKitButton.js
import React from "react";
import { TertiaryButton } from "@workday/canvas-kit-react/button";
import { chevronUpSmallIcon } from "@workday/canvas-system-icons-web";

export default () => (
  <TertiaryButton
    icon={chevronUpSmallIcon}
    iconPosition="right"
    onClick={() => {}}
    size="small"
  >
    CanvasKitButton
  </TertiaryButton>
);

// CanvasKitButtonWithoutIcon.js
import React from "react";
import { TertiaryButton } from "@workday/canvas-kit-react/button";

export default () => (
  <TertiaryButton onClick={() => {}} size="small">
    CanvasKitButton
  </TertiaryButton>
);


// Test code
import "@testing-library/jest-dom/extend-expect";
import React from "react";
import { render } from "@testing-library/react";
import CanvasKitButton from "../CanvasKitButton";
import CanvasKitButtonWithoutIcon from "../CanvasKitButtonWithoutIcon";

test("This fails with TypeError", () => {
  const { getByRole } = render(<CanvasKitButton />);
  expect(getByRole("button", { name: "CanvasKitButton" })).toBeInTheDocument();
});

test("This passes", () => {
  const { getByRole } = render(<CanvasKitButtonWithoutIcon />);
  expect(getByRole("button", { name: "CanvasKitButton" })).toBeInTheDocument();
});

```

### What you did:

When trying to use `getByRole` queries in my application using Canvas Kit, I found that if I had buttons on my page with icons, the `getByRole` queries were borking out with a very strange error: "TypeError: Cannot read property 'parentElement' of null". 

When I remove the icon or mock the underlying icon library, the test passes.

I found a similar error here but no solution other than to change to a different query: https://spectrum.chat/testing-library/help-react/getbyrole-error-cannot-read-property-parentelement-of-null~0b78b6db-b99f-414e-b9ed-38990d6dd717

### What happened:

Repo error:

```
 FAIL  src/CanvasKitButton.test.js
  ✕ This fails with TypeError (177 ms)
  ✓ This should pass (108 ms)

    TypeError: Cannot read property 'parentElement' of null

      at Array.Resolver (eval at compile (node_modules/nwsapi/src/nwsapi.js:768:17), <anonymous>:3:140)
      at match_assert (node_modules/nwsapi/src/nwsapi.js:1351:13)
      at Object._matches [as match] (node_modules/nwsapi/src/nwsapi.js:1369:16)
      at exports.matchesDontThrow (node_modules/jsdom/lib/jsdom/living/helpers/selectors.js:29:36)
      at matches (node_modules/jsdom/lib/jsdom/living/helpers/style-rules.js:50:10)
      at node_modules/jsdom/lib/jsdom/living/helpers/style-rules.js:35:18
          at Array.forEach (<anonymous>)
      at handleSheet (node_modules/jsdom/lib/jsdom/living/helpers/style-rules.js:26:13)
          at Array.forEach (<anonymous>)
      at exports.forEachMatchingSheetRuleOfElement (node_modules/jsdom/lib/jsdom/living/helpers/style-rules.js:46:11)
```

Codesandbox error:
```
Cannot read properties of null (reading 'parentElement')

TypeError: Cannot read properties of null (reading 'parentElement')
    at Array.Resolver (eval at compile (https://un6lk.csb.app/static/js/jsdom-16.3.0.min.js:236:17232), <anonymous>:3:140)
    at match_assert (https://un6lk.csb.app/static/js/jsdom-16.3.0.min.js:236:30554)
    at Object._matches [as match] (https://un6lk.csb.app/static/js/jsdom-16.3.0.min.js:236:30873)
    at exports.matchesDontThrow (https://un6lk.csb.app/static/js/jsdom-16.3.0.min.js:37:1612417)
    at matches (https://un6lk.csb.app/static/js/jsdom-16.3.0.min.js:37:1620955)
    at https://un6lk.csb.app/static/js/jsdom-16.3.0.min.js:37:1620677
    at Array.forEach (<anonymous>)
    at handleSheet (https://un6lk.csb.app/static/js/jsdom-16.3.0.min.js:37:1620484)
    at Array.forEach (<anonymous>)
    at exports.forEachMatchingSheetRuleOfElement (https://un6lk.csb.app/static/js/jsdom-16.3.0.min.js:37:1620853)
    at Window.getComputedStyle (https://un6lk.csb.app/static/js/jsdom-16.3.0.min.js:37:49816)
    at isHidden (https://un6lk.csb.app/node_modules/dom-accessibility-api/dist/accessible-name.js:66:56)
    at computeTextAlternative (https://un6lk.csb.app/node_modules/dom-accessibility-api/dist/accessible-name.js:314:13)
    at computeMiscTextAlternative (https://un6lk.csb.app/node_modules/dom-accessibility-api/dist/accessible-name.js:231:28)
    at computeTextAlternative (https://un6lk.csb.app/node_modules/dom-accessibility-api/dist/accessible-name.js:413:20)
    at computeMiscTextAlternative (https://un6lk.csb.app/node_modules/dom-accessibility-api/dist/accessible-name.js:231:28)
    at computeTextAlternative (https://un6lk.csb.app/node_modules/dom-accessibility-api/dist/accessible-name.js:413:20)
    at computeMiscTextAlternative (https://un6lk.csb.app/node_modules/dom-accessibility-api/dist/accessible-name.js:231:28)
    at computeTextAlternative (https://un6lk.csb.app/node_modules/dom-accessibility-api/dist/accessible-name.js:413:20)
    at computeMiscTextAlternative (https://un6lk.csb.app/node_modules/dom-accessibility-api/dist/accessible-name.js:231:28)
    at computeTextAlternative (https://un6lk.csb.app/node_modules/dom-accessibility-api/dist/accessible-name.js:413:20)
    at computeMiscTextAlternative (https://un6lk.csb.app/node_modules/dom-accessibility-api/dist/accessible-name.js:231:28)
    at computeTextAlternative (https://un6lk.csb.app/node_modules/dom-accessibility-api/dist/accessible-name.js:402:20)
    at computeAccessibleName (https://un6lk.csb.app/node_modules/dom-accessibility-api/dist/accessible-name.js:427:25)
    at eval (https://un6lk.csb.app/node_modules/@testing-library/dom/dist/@testing-library/dom.esm.js:1190:68)
    at Array.filter (<anonymous>)
    at queryAllByRole (https://un6lk.csb.app/node_modules/@testing-library/dom/dist/@testing-library/dom.esm.js:1186:6)
    at eval (https://un6lk.csb.app/node_modules/@testing-library/dom/dist/@testing-library/dom.esm.js:795:24)
    at eval (https://un6lk.csb.app/node_modules/@testing-library/dom/dist/@testing-library/dom.esm.js:783:24)
    at Object.eval (https://un6lk.csb.app/src/__tests__/CanvasKitButton.test.js:26:10)
    at https://codesandbox.io/static/js/3.c68bd71c0.chunk.js:1:336366
    at new Promise (<anonymous>)
    at t.callAsyncFn (https://codesandbox.io/static/js/3.c68bd71c0.chunk.js:1:336025)
    at https://codesandbox.io/static/js/7.cdfc7986d.chunk.js:1:6524
    at c (https://codesandbox.io/static/js/vendors~app~embed~sandbox~sandbox-startup.bcc15d438.chunk.js:1:3629)
    at Generator._invoke (https://codesandbox.io/static/js/vendors~app~embed~sandbox~sandbox-startup.bcc15d438.chunk.js:1:3382)
    at Generator.forEach.t.<computed> [as next] (https://codesandbox.io/static/js/vendors~app~embed~sandbox~sandbox-startup.bcc15d438.chunk.js:1:3986)
    at r (https://codesandbox.io/static/js/vendors~app~embed~sandbox~sandbox-startup.bcc15d438.chunk.js:1:206)
    at u (https://codesandbox.io/static/js/vendors~app~embed~sandbox~sandbox-startup.bcc15d438.chunk.js:1:417)
    at https://codesandbox.io/static/js/vendors~app~embed~sandbox~sandbox-startup.bcc15d438.chunk.js:1:476
    at new Promise (<anonymous>)
    at https://codesandbox.io/static/js/vendors~app~embed~sandbox~sandbox-startup.bcc15d438.chunk.js:1:357
    at https://codesandbox.io/static/js/7.cdfc7986d.chunk.js:1:6791
    at https://codesandbox.io/static/js/7.cdfc7986d.chunk.js:1:5456
    at c (https://codesandbox.io/static/js/vendors~app~embed~sandbox~sandbox-startup.bcc15d438.chunk.js:1:3629)
    at Generator._invoke (https://codesandbox.io/static/js/vendors~app~embed~sandbox~sandbox-startup.bcc15d438.chunk.js:1:3382)
    at Generator.forEach.t.<computed> [as next] (https://codesandbox.io/static/js/vendors~app~embed~sandbox~sandbox-startup.bcc15d438.chunk.js:1:3986)
    at r (https://codesandbox.io/static/js/vendors~app~embed~sandbox~sandbox-startup.bcc15d438.chunk.js:1:206)
    at u (https://codesandbox.io/static/js/vendors~app~embed~sandbox~sandbox-startup.bcc15d438.chunk.js:1:417)
    at https://codesandbox.io/static/js/vendors~app~embed~sandbox~sandbox-startup.bcc15d438.chunk.js:1:476
```

### Reproduction:

Github repository:
https://github.com/ahayes91/canvas-kit-react-testing-library/blob/main/src/CanvasKitButton.test.js

Codesandbox
[![Edit react-testing-library demo (forked)](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/react-testing-library-demo-forked-un6lk?fontsize=14&hidenavigation=1&theme=dark)

### Problem description:

- getByRole should pass regardless of whether an icon is provided on the button or not
- Cryptic errors make it hard to debug if it's your code or the test that is wrong
- New engineers using @testing-library/react are put off by scary errors 😂 

### Suggested solution:

I don't have one I'm afraid, but when I went digging into the stack trace, I changed the nswapi logging in my local node modules `node_modules/jsdom/lib/jsdom/living/helpers/selectors.js` file:

```
exports.matchesDontThrow = (elImpl, selector) => {
  const document = elImpl._ownerDocument;

  if (!document._nwsapiDontThrow) {
    document._nwsapiDontThrow = initNwsapi(elImpl);
    document._nwsapiDontThrow.configure({
      LOGERRORS: true,
      VERBOSITY: true,
      IDS_DUPES: true,
      MIXEDCASE: true
    });
  }

  return document._nwsapiDontThrow.match(selector, idlUtils.wrapperForImpl(elImpl));
};
```

Which gave this error when I reran the tests, which was interesting but then I found myself pretty stumped without better knowledge of the underlying libraries:

```
 FAIL  src/CanvasKitButton.test.js
  ✕ This fails with TypeError (77 ms)
  ✕ This should pass (39 ms)

  ● This fails with TypeError

    SyntaxError: '@namespace "http://www.w3.org/1999/xhtml";

    html' is not a valid selector

      at emit (node_modules/nwsapi/src/nwsapi.js:565:17)
      at Object._matches [as match] (node_modules/nwsapi/src/nwsapi.js:1409:9)
      at exports.matchesDontThrow (node_modules/jsdom/lib/jsdom/living/helpers/selectors.js:29:36)
      at matches (node_modules/jsdom/lib/jsdom/living/helpers/style-rules.js:50:10)
      at node_modules/jsdom/lib/jsdom/living/helpers/style-rules.js:35:18
          at Array.forEach (<anonymous>)
      at handleSheet (node_modules/jsdom/lib/jsdom/living/helpers/style-rules.js:26:13)
      at exports.forEachMatchingSheetRuleOfElement (node_modules/jsdom/lib/jsdom/living/helpers/style-rules.js:45:3)
      at getComputedStyleImplementation (node_modules/jsdom/lib/jsdom/browser/Window.js:802:5)
      at isHidden (node_modules/dom-accessibility-api/sources/accessible-name-and-description.ts:84:16)

  ● This should pass

    SyntaxError: '@namespace "http://www.w3.org/1999/xhtml";

    html' is not a valid selector

      at emit (node_modules/nwsapi/src/nwsapi.js:565:17)
      at Object._matches [as match] (node_modules/nwsapi/src/nwsapi.js:1409:9)
      at exports.matchesDontThrow (node_modules/jsdom/lib/jsdom/living/helpers/selectors.js:29:36)
      at matches (node_modules/jsdom/lib/jsdom/living/helpers/style-rules.js:50:10)
      at node_modules/jsdom/lib/jsdom/living/helpers/style-rules.js:35:18
          at Array.forEach (<anonymous>)
      at handleSheet (node_modules/jsdom/lib/jsdom/living/helpers/style-rules.js:26:13)
      at exports.forEachMatchingSheetRuleOfElement (node_modules/jsdom/lib/jsdom/living/helpers/style-rules.js:45:3)
      at getComputedStyleImplementation (node_modules/jsdom/lib/jsdom/browser/Window.js:802:5)
      at isHidden (node_modules/dom-accessibility-api/sources/accessible-name-and-description.ts:84:16)
```

**Comment by ahayes91** [comment]
Wee update on this - you might see the test that _should_ pass failing in the code sandbox above (and I completely missed it when I wrote out the results too) with an error about not being able to find any accessible roles. 
Updating to @testing-library/react 12.1.4 resolves that issue on the sandbox - @NicholasBoll astutely noticed that there were some recent changes in [`dom-accessibility-api`](https://github.com/eps1lon/dom-accessibility-api/releases) that may have contributed to this re. calculating the name of inaccessible elements.

Worth noting that updating @testing-library/react or even pinning `dom-accessibility-api:"0.5.13"` as a dependency in the local repo doesn't fix the `TypeError: Cannot read property 'parentElement' of null` error.

**Comment by tkall-newsela** [comment]
Not sure if related but I'm having this error `TypeError: Cannot read property 'getPropertyValue' of undefined` when I try to use `screen.getByRole('button', { ... });`

**Comment by willklein** [comment]
@tpict Hey, I looked into this further. We actually still have the problem testing a component in our library, Canvas Kit, even on the latest version. I debugged this and dug deeper and couldn't reproduce it initially with a non-Canvas Kit element, even replicating some styles, but not all. So I dove into the libraries in play and mapped the function calls to the stack trace, and I think this is the issue that's cropping up: https://github.com/jsdom/jsdom/issues/3057

If so, this won't be a `dom-testing-library` issue or a `@workday/canvas-kit-react` issue, but something in `jsdom` most likely.

I'm going to follow this rabbit hole into that utility and see what I can figure out! They had all the obvious null checks for `parentElement` so it might not be easy to solve, but I'll keep digging.

**Comment by tpict** [comment]
@willklein Funnily enough I've been suffering the exact same issue when `jsdom` attempts to parse CSS selectors that target SVG elements. Thanks for the update!

**Comment by tpict** [comment]
@willklein Thanks for the update!

**Comment by ahayes91** [comment]
I have some more interesting findings today - it looks like when I update the buttons with icons to use the correct `iconPosition` prop value, they start failing again.

<iframe src="https://codesandbox.io/embed/react-testing-library-fireevent-click-demo-with-ck-7-un6lk?fontsize=14&hidenavigation=1&theme=dark"
     style="width:100%; height:500px; border:0; border-radius: 4px; overflow:hidden;"
     title="react-testing-library fireEvent.click demo with CK 7"
     allow="accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking"
     sandbox="allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"
   ></iframe>

https://github.com/ahayes91/canvas-kit-react-testing-library/blob/main/src/__tests__/ButtonsIconsAndTooltips.test.js has also been updated to match the code sandbox and is showing the same results locally.

**Comment by phoenisx** [comment]
This issue does not seem to be a `dom-testing-library` issue. It's related to the underlying `nwsapi` library that RTL uses to query CSS selectors.

A PR to fix this issue has been raised here: https://github.com/dperini/nwsapi/pull/78, which possibly would resolve this issue as well.

**Comment by eps1lon** [comment]
If this issue persists, please run a `yarn why nwsapi` or `npm ls nwsapi` and post the result. This was supposedly fixed in `nwsapi@2.2.7`

**Comment by ahayes91** [comment]
Sorted! Just updated and tested https://github.com/ahayes91/canvas-kit-react-testing-library/blob/main/src/__tests__/ButtonsIconsAndTooltips.test.js again and it all passes now with the latest Canvas Kit and Testing Library versions 🎉 
Thank you!
